### PR TITLE
pool flate writers

### PR DESCRIPTION
**Pull request by y3llowcake**

### CODE REVIEW

**Comment by y3llowcake** [comment]
Hoping this is relatively uncontroversial, in theory it should only improve on existing performance characteristics.

File: compression.go [hunk 1]
```diff
@@ -73,6 +85,13 @@ func (w *flateWrapper) Write(p []byte) (int, error) {
...
 func (w *flateWrapper) Close() error {
+	if w.fw == nil {
+		return errors.New("already closed")
+	}
+	defer func() {
+		flateWriterPool.Put(w.fw)
+		w.fw = nil
+	}()
 	err1 := w.fw.Flush()
 	if w.tw.p != [4]byte{0, 0, 0xff, 0xff} {
```
Comment (garyburd) [inline compression.go]: Remove the defer and add the following after Flush()

     flatWritePool.Put(w.fw)
     w.fw = nil

File: compression.go [hunk 2]
```diff
@@ -73,6 +85,13 @@ func (w *flateWrapper) Write(p []byte) (int, error) {
 }
 
 func (w *flateWrapper) Close() error {
+	if w.fw == nil {
+		return errors.New("already closed")
```
Comment (garyburd) [inline compression.go]: Return [errWriteClosed](https://github.com/gorilla/websocket/blob/0868951cdb8e69bc42df4598bdc6164ff2f1a072/conn.go#L179).

**Comment by y3llowcake** [comment]
Done

**Comment by garyburd** [comment]
Does not compile.

**Comment by y3llowcake** [comment]
Sorry, I'm terrible with computers. This should be fixed now.

**Pull request merged**